File: CSharpSegmenter/Segment.cs

using System;
using System.ComponentModel.Design;
using System.Runtime.CompilerServices;

namespace CSharpSegmenter
{
    // OO representation of a Segment as an array of child pixels.
    // Contains a cached value
    public class Segment
    {
        public Pixel[] pixels { get; private set; }
        // cache for standard deviation calculation
        public double[] stdDevs { get; private set; }
        
        // Segment constructor for the bottom level of the Segmentation
        // A segment with a single pixel only
        public Segment(Pixel pixel)
        {
            this.pixels = new Pixel[1] {pixel};
        }

        // Constructor for creating a parent Segment from two children
        public Segment(Segment segment1, Segment segment2)
        {
            int length1 = segment1.pixels.Length;
            int length2 = segment2.pixels.Length;

            this.pixels = new Pixel[length1 + length2];
            segment1.pixels.CopyTo(this.pixels, 0);
            segment2.pixels.CopyTo(this.pixels, length1);
        }
        
        // determine the cost of merging the given segments in our Segmentation algorithm: 
        // equal to the standard deviation of the combined the segments minus the sum
        // of the standard deviations of the individual segments, weighted by their
        // respective sizes and summed over all colour bands
        public static double MergeCost(Segment segment1, Segment segment2)
        {
            double combinedWeight = CalculateSegmentWeight(new Segment(segment1, segment2));
            double segment1Weight = CalculateSegmentWeight(segment1);
            double segment2Weight = CalculateSegmentWeight(segment2);

            return combinedWeight - segment1Weight - segment2Weight;
        }
        
        // Calculate the contribution of a specific segment to the merge cost of two segments
        private static double CalculateSegmentWeight(Segment segment)
        {
            segment.CalculateStdDevs();   
            
            double sum = 0;
            foreach (var deviation in segment.stdDevs) sum += deviation;
            return sum * segment.pixels.Length;
        }
        
        // Return a list of the standard deviations of the pixel colours in the given segment.
        // The list contains one entry for each colour band, typically: [red, green and blue].
        private double[] CalculateStdDevs()
        {
            // return a cached result if possible
            if (stdDevs != null) return stdDevs;
            
            // Get dimensions
            int numBands = GetNumSubPixelsPerPixel();
            int numPixels = pixels.Length;
            
            // An array of all subpixels in a band for a Segment
            // [ [r0; r1; r2; r3; ... rN]; [g0; g1; g2; g3; ... gN]; [b0; b1; b2; b3; ... bN] ]
            byte[][] subPixels = Transpose(pixels, numPixels, numBands);
            
            // Std dev for each subpixel band
            stdDevs = new double[numBands];
            for (int i=0; i<numBands; i++) this.stdDevs[i] = stdDev(subPixels[i]);
            return stdDevs;
        }

        private byte[][] Transpose(Pixel[] arr, int numPixels, int numBands)
        {
            byte[][] subPixels = new byte[numBands][];
            for (int j=0; j<numBands; j++)
            {
                subPixels[j] = new byte[numPixels];
                for (int i=0; i<numPixels; i++) subPixels[j][i] = pixels[i][j];
            }
            return subPixels;
        }

        // Returns the number of subpixels per pixel for an image.
        // There should be no cases in this program where a pixel object is not allocated,
        // but we access the length cautiously as it is an infrequest operation anyway.
        private int GetNumSubPixelsPerPixel()
        {
            if (pixels == null) 
                throw new NullReferenceException("Cannot get num subpixels per pixel. Segment does not contain pixels.");
            if (pixels[0] == null) throw new 
                IndexOutOfRangeException("Cannot get num subpixels per pixel. First pixel missing from Segment");
            return pixels[0].subPixels.Length;
        }

        private double stdDev(byte[] arr)
        {
            double sum = 0;
            foreach (var x in arr) sum += x;
            double average = sum / arr.Length;

            double sumOfSquareMeanDistances = 0;
            foreach (var x in arr) sumOfSquareMeanDistances += (x - average) * (x - average);

            double variance = sumOfSquareMeanDistances / arr.Length;
            return (double) Math.Sqrt(variance);
        }

        public override bool Equals(object obj)
        {
            // Check for null values and compare run-time types.
            if (obj == null || this.GetType() != obj.GetType()) 
                return false;

            Segment other = (Segment) obj;
            
            
            if (this.pixels.Length != other.pixels.Length)
                return false;
            
            // expects pixels to be in the same array order
            for (int i = 0; i < this.pixels.Length; i++)
            {
                if ( !(this.pixels[i] == (other.pixels[i])) )
                    return false;
            }

            return true;
        }
    }
}

File: FSharpImpureSegmenter/Segment.fs
module Segment

open System
open System.Collections.Immutable
open Util

type Coord = { x: int; y: int } 

type Pixel = { x: int; y: int; subPixels: byte array }

// Project a Pixel down to just a Coord
let getCoordFromPixel (pixel: Pixel) : Coord = 
    {x = pixel.x; y = pixel.x}

// A Segment is just a collection of pixels
type Segment = Pixel array

// Default Segment constructor
let Segment (pixel: Pixel) : Segment = [| pixel |] 

// Alternative Segment constructor
let createSegment (segment1: Segment) (segment2: Segment) : Segment = 
    Array.concat [segment1; segment2]
    

// Calculates the statistical standard deviation of a vector of numbers
let standardDeviation (elems: float array) : float = 
    elems
    |> Array.map (fun x -> x - (Array.average elems) )
    |> Array.map square
    |> Array.sum
    |> divideBy (float(Array.length elems))
    |> sqrt


// return a vector of the standard deviations of the pixel colours in the given segment
// the list contains one entry for each colour band, typically: [red, green and blue]
let stddev (segment: Segment) = 
    let numPixels = segment.Length
    let numBands = (Array.last segment).subPixels.Length
    
    // get all subpixels bytes, grouped by subpixel, expected structure:
    // [ [r0; r1; r2; r3; ... rN]; [g0; g1; g2; g3; ... gN]; [b0; b1; b2; b3; ... bN] ]
    let subPixelBands: byte array array = init2d numPixels numBands (fun (x: int) (y: int) -> segment.[y].subPixels.[x])

    let stdDevBands =
        subPixelBands
        |> Array.map (Array.map (fun x -> float x) ) // convert byte to float
        |> Array.map standardDeviation
    
    stdDevBands


// Calculate the contribution of a segment to the merge cost of two segments
let calculateSegmentWeight (segment: Segment) : float =
    let subPixelStdDevs: float array = stddev(segment)
    
    subPixelStdDevs
    |> Array.sum
    |> multiplyBy (float(segment.Length))


// determine the cost of merging the given segments: 
// equal to the standard deviation of the combined the segments minus the sum of the standard deviations of the individual segments,
// weighted by their respective sizes and summed over all colour bands
let mergeCost (segment1: Segment) (segment2: Segment) : float =
    let combinedWeight: float  = calculateSegmentWeight(createSegment segment1 segment2)
    let segment1Weight: float  = calculateSegmentWeight segment1
    let segment2Weight: float  = calculateSegmentWeight segment2

    combinedWeight - segment1Weight - segment2Weight;
